### PR TITLE
Use TYPE_CHECKING special constant

### PR DESCRIPTION
**Pull request by PeterJCLaw**
It turns out this was introduced in Python 3.5.2, much earlier than we'd thought so we can make use of it.

### CODE REVIEW

**Comment by danpalmer** [comment]
Approved in principle, but it looks like the failure is due to the layering tests failing, which could indicate that the layering tests are over eagerly chasing the typechecking imports.

**Comment by PeterJCLaw** [comment]
Indeed, I _think_ that this is a consequence of the change from a literal constant to a variable (which happens to be invariant).

I'm not sure whether the correct resolution would be to adjust the layering test to ignore typing imports or to adjust the imports we allow; that comes down to what we're trying to achieve with them.

**Comment by prophile** [comment]
I feel like the test failure here is correct: requiring a `Context` subclass instance in typing _is_ an explicit dependency on the `context` component which the layering test is uncovering.

**Comment by PeterJCLaw** [comment]
Huh, so we can't _just_ fix this as that results in a circular dependency, which `test_layering` then also complains about. The circularity is actually fine as far as Python is concerned though.

**Comment by prophile** [comment]
We should either:

1. Move the common component out to a separate model depended on by both components in the cycle, or
2. Factor out `Context` so we parameterise by a Context-like object in a `TypeVar`.

File: routemaster/exit_conditions/program.py [hunk 1]
```diff
@@ -11,7 +11,7 @@
     format_parse_error_message,
 )
 
-if False:  # typing
+if typing.TYPE_CHECKING:
```
Comment (danpalmer) [inline routemaster/exit_conditions/program.py]: Do you mind splitting out all the imports from the typing module? That's how we do it elsewhere in the codebase so would be good to be consistent.

Comment (PeterJCLaw) [reply routemaster/exit_conditions/program.py]: Sure, I'll put up a separate PR for that after merging this.

**Pull request merged**